File: stepik-clone/frontend/lib/features/courses/domain/models/course.dart
// frontend/lib/features/courses/domain/models/course.dart
class Module {
  final int id;
  final String title;
  final int order;

  Module({
    required this.id,
    required this.title,
    required this.order,
  });

  factory Module.fromJson(Map<String, dynamic> json) {
    return Module(
      id: json['id'],
      title: json['title'],
      order: json['order'],
    );
  }
}

class Course {
  final int id;
  final String title;
  final String? description;
  final String? coverImageUrl;
  final int teacherId;
  final bool isPaid;
  final double price;
  final bool isPublished;
  final List<Module> modules;

  Course({
    required this.id,
    required this.title,
    this.description,
    this.coverImageUrl,
    required this.teacherId,
    required this.isPaid,
    required this.price,
    required this.isPublished,
    this.modules = const [],
  });

  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      coverImageUrl: json['cover_image_url'],
      teacherId: json['teacher_id'],
      isPaid: json['is_paid'] ?? false,
      price: (json['price'] ?? 0.0).toDouble(),
      isPublished: json['is_published'] ?? false,
      modules: (json['modules'] as List?)
              ?.map((e) => Module.fromJson(e))
              .toList() ??
          const [],
    );
  }
}

File: stepik-clone/backend/app/models/course.py
# backend/app/models/course.py
import enum
from sqlalchemy import Column, Integer, String, Boolean, Float, ForeignKey, DateTime, Enum, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database import Base

class Category(Base):
    __tablename__ = "categories"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, unique=True, index=True, nullable=False)
    description = Column(String)

    courses = relationship("Course", back_populates="category")

class Course(Base):
    __tablename__ = "courses"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, index=True, nullable=False)
    description = Column(String)
    cover_image_url = Column(String)
    category_id = Column(Integer, ForeignKey("categories.id"))
    teacher_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    is_paid = Column(Boolean, default=False)
    price = Column(Float, default=0.0)
    is_published = Column(Boolean, default=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    category = relationship("Category", back_populates="courses")
    modules = relationship("Module", back_populates="course", cascade="all, delete-orphan")

class Module(Base):
    __tablename__ = "modules"

    id = Column(Integer, primary_key=True, index=True)
    course_id = Column(Integer, ForeignKey("courses.id"), nullable=False)
    title = Column(String, nullable=False)
    order = Column(Integer, nullable=False)

    course = relationship("Course", back_populates="modules")
    lessons = relationship("Lesson", back_populates="module", cascade="all, delete-orphan")

class LessonTypeEnum(str, enum.Enum):
    video = "video"
    text = "text"
    quiz = "quiz"
    code = "code"

class Lesson(Base):
    __tablename__ = "lessons"

    id = Column(Integer, primary_key=True, index=True)
    module_id = Column(Integer, ForeignKey("modules.id"), nullable=False)
    title = Column(String, nullable=False)
    order = Column(Integer, nullable=False)
    type = Column(Enum(LessonTypeEnum), nullable=False)

    module = relationship("Module", back_populates="lessons")
    content = relationship("LessonContent", back_populates="lesson", uselist=False, cascade="all, delete-orphan")
    quiz = relationship("Quiz", back_populates="lesson", uselist=False, cascade="all, delete-orphan")
